x-pack/plugin/data-streams/src/main/java/org/elasticsearch/xpack/datastreams/UpdateTimeSeriesRangeService.java

```diff
@@ -0,0 +1,170 @@
+                + DataStreamsPlugin.TIME_SERIES_POLL_INTERVAL.getKey()
+                + "] setting from ["
+                + pollInterval
+                + "] to ["
+                + newValue
+                + "]"
+        );
+        this.pollInterval = newValue;
+    }
+
+    ClusterState updateTimeSeriesTemporalRange(ClusterState current, Instant now) {
+        Metadata.Builder mBuilder = null;
+        for (DataStream dataStream : current.metadata().dataStreams().values()) {
+            if (dataStream.getIndexMode() != IndexMode.TIME_SERIES) {
+                continue;
+            }
+            if (dataStream.isReplicated()) {
+                continue;
+            }
+
+            // getWriteIndex() selects the latest added index:
+            String head = dataStream.getWriteIndex().getName();
+            IndexMetadata im = current.metadata().getIndices().get(head);
+            Instant currentEnd = IndexSettings.TIME_SERIES_END_TIME.get(im.getSettings());
+            TimeValue lookAheadTime = IndexSettings.LOOK_AHEAD_TIME.get(im.getSettings());
+            Instant newEnd = now.plus(lookAheadTime.getMillis(), ChronoUnit.MILLIS).plus(pollInterval.getMillis(), ChronoUnit.MILLIS);
+            if (newEnd.isAfter(currentEnd)) {
+                Settings settings = Settings.builder()
+                    .put(IndexSettings.TIME_SERIES_END_TIME.getKey(), DEFAULT_DATE_TIME_FORMATTER.format(newEnd))
+                    .build();
+                LOGGER.debug(
+                    "updating [{}] setting from [{}] to [{}] for index [{}]",
+                    IndexSettings.TIME_SERIES_END_TIME.getKey(),
+                    currentEnd,
+                    newEnd,
+                    head
+                );
+                if (mBuilder == null) {
+                    mBuilder = Metadata.builder(current.metadata());
+                }
+                mBuilder.updateSettings(settings, head);
+                Metadata.Builder finalMBuilder = mBuilder;
+                // Verify that all temporal ranges of each backing index is still valid:
+                dataStream.validate(finalMBuilder::get);
+            }
+        }
+
+        if (mBuilder != null) {
+            return ClusterState.builder(current).metadata(mBuilder).build();
+        } else {
+            return current;
+        }
+    }
+
+    void scheduleTask() {
+        if (job == null) {
+            LOGGER.debug("schedule tsdb update task");
+            job = threadPool.schedule(() -> perform(this::scheduleTask), pollInterval, ThreadPool.Names.SAME);
+        }
+    }
+
+    void unschedule() {
+        if (job != null) {
+            job.cancel();
+            job = null;
+        }
+    }
+
+    @Override
+    protected void doStart() {
+        clusterService.addLocalNodeMasterListener(this);
+    }
+
+    @Override
+    protected void doStop() {
+        unschedule();
+    }
+
+    @Override
+    protected void doClose() throws IOException {}
```

Comment:
I think we should call `unschedule()` in the close method here also, so that it doesn't try to execute during shutdown.